### PR TITLE
3D NoC

### PR DESCRIPTION
**Pull request by soheilshahrouz**
#### Description
Generalizing the NoC data structures, architecture description and routing algorithms to 3D (multi-layer) FPGAs.

#### Motivation and Context
VTR is mostly 3D now, but the NoCs are still only 2D.

#### How Has This Been Tested?
Created a new arch file and ran synthetic benchmarks on this new 3D NoC arch within the 3D Stratix IV arch. The 10x10 2D mesh is replaced by two 7x7 meshes stacked on top of each other. All the benchmarks passed, and the QoR metrics are slightly better as the 3D NoC topology shortens some traffic flows.

CI has been updated to run some tests on this 3D architecture.

#### Types of changes
- [x] New feature (change which adds functionality)

#### Checklist:
- [x] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by vaughnbetz** [comment]
Some CI test failures due to vtr::isclose getting one float and one double; you'll have to explicitly cast or change a constant or some such (@soheilshahrouz tells me he has float constants and double variables so that is probably it).

**Comment by vaughnbetz** [comment]
Until CI is sorted please run the major nightly tests on wintermute and list the results here.

File: vpr/src/noc/noc_traffic_flows.h [hunk 1]
```diff
@@ -237,6 +237,8 @@ class NocTrafficFlows {
      */
     const std::vector<NocTrafficFlowId>& get_all_traffic_flow_id() const;
 
+    const vtr::vector<NocTrafficFlowId, t_noc_traffic_flow>& get_all_traffic_flows() const;
```
Comment (vaughnbetz) [inline vpr/src/noc/noc_traffic_flows.h:240]: Doxygen comment.

File: vpr/src/noc/odd_even_routing.h [hunk 2]
```diff
@@ -51,6 +58,9 @@ class OddEvenRouting : public TurnModelRouting{
      * @return True if the passed number of even, otherwise false.
      */
     static inline bool is_even(int number);
+
+  private:
+    vtr::vector<NocRouterId, t_physical_tile_loc> compressed_noc_locs_;
```
Comment (vaughnbetz) [inline vpr/src/noc/odd_even_routing.h:63]: Doxygen comment of what this is.

File: vpr/src/noc/odd_even_routing.h [hunk 3]
```diff
@@ -18,25 +18,32 @@
...
-    TurnModelRouting::Direction select_next_direction(const std::vector<TurnModelRouting::Direction>& legal_directions,
-                                                      NocRouterId src_router_id,
-                                                      NocRouterId dst_router_id,
-                                                      NocRouterId curr_router_id,
-                                                      NocTrafficFlowId traffic_flow_id,
-                                                      const NocStorage& noc_model) override;
+    bool is_turn_legal(const std::array<std::reference_wrapper<const NocRouter>, 3>& noc_routers,
+                       const NocStorage& noc_model) const override;
+
+    inline void route_2d(t_physical_tile_loc comp_src_loc,
```
Comment (vaughnbetz) [inline vpr/src/noc/odd_even_routing.h:38]: Doxygen comment

File: vpr/src/noc/odd_even_routing.h [hunk 4]
```diff
@@ -18,25 +18,32 @@
...
+                         t_physical_tile_loc comp_curr_loc,
+                         t_physical_tile_loc comp_dst_loc,
+                         TurnModelRouting::Direction prev_dir);
 
-    bool is_turn_legal(const std::array<std::reference_wrapper<const NocRouter>, 3>& noc_routers) const override;
+    inline void route_3d(t_physical_tile_loc comp_src_loc,
```
Comment (vaughnbetz) [inline vpr/src/noc/odd_even_routing.h:43]: Doxygen comment

File: vpr/src/noc/odd_even_routing.h [hunk 5]
```diff
@@ -18,25 +18,32 @@
...
+                                                                         TurnModelRouting::Direction prev_dir,
                                                                          const NocStorage& noc_model) override;
 
-    TurnModelRouting::Direction select_next_direction(const std::vector<TurnModelRouting::Direction>& legal_directions,
-                                                      NocRouterId src_router_id,
-                                                      NocRouterId dst_router_id,
-                                                      NocRouterId curr_router_id,
-                                                      NocTrafficFlowId traffic_flow_id,
-                                                      const NocStorage& noc_model) override;
+    bool is_turn_legal(const std::array<std::reference_wrapper<const NocRouter>, 3>& noc_routers,
```
Comment (vaughnbetz) [inline vpr/src/noc/odd_even_routing.h:35]: Doxygen comment

File: vpr/src/noc/odd_even_routing.cpp [hunk 6]
```diff
@@ -1,152 +1,75 @@
...
+    std::set<int> unique_x, unique_y, unique_z;
+
+    for (const NocRouter& router : noc_model.get_noc_routers()) {
+        t_physical_tile_loc loc = router.get_router_physical_location();
+        unique_x.insert(loc.x);
+        unique_y.insert(loc.y);
+        unique_z.insert(loc.layer_num);
+    }
+
+    auto create_compressed_map = [](const std::set<int> unique_vals) -> std::map<int, int> {
```
Comment (vaughnbetz) [inline vpr/src/noc/odd_even_routing.cpp:19]: pass set by const ref

File: vpr/src/noc/odd_even_routing.cpp [hunk 7]
```diff
@@ -1,152 +1,75 @@
...
+OddEvenRouting::OddEvenRouting(const NocStorage& noc_model) {
+    std::set<int> unique_x, unique_y, unique_z;
+
+    for (const NocRouter& router : noc_model.get_noc_routers()) {
+        t_physical_tile_loc loc = router.get_router_physical_location();
+        unique_x.insert(loc.x);
+        unique_y.insert(loc.y);
+        unique_z.insert(loc.layer_num);
+    }
+
```
Comment (vaughnbetz) [inline vpr/src/noc/odd_even_routing.cpp:18]: Comment what this does --> creates a map with each entry in the passed in set having a value assigned that is a unique integer starting at 0

File: vpr/src/noc/odd_even_routing.cpp [hunk 8]
```diff
@@ -1,152 +1,75 @@
...
+    std::map<int, int> compressed_y_map = create_compressed_map(unique_y);
+    std::map<int, int> compressed_z_map = create_compressed_map(unique_z);
+
+    compressed_noc_locs_.resize(noc_model.get_number_of_noc_routers());
+
+    for (const auto& [id, router] : noc_model.get_noc_routers().pairs()) {
+        t_physical_tile_loc loc = router.get_router_physical_location();
+
+        t_physical_tile_loc compressed_loc{compressed_x_map[loc.x],
+                                           compressed_x_map[loc.y],
```
Comment (vaughnbetz) [inline vpr/src/noc/odd_even_routing.cpp:39]: compressed_y_map

File: vpr/src/noc/odd_even_routing.cpp [hunk 9]
```diff
@@ -1,152 +1,75 @@
...
+                                           compressed_x_map[loc.layer_num]};
```
Comment (vaughnbetz) [inline vpr/src/noc/odd_even_routing.cpp:40]: compressed_z_map

File: vpr/src/noc/odd_even_routing.cpp [hunk 10]
```diff
@@ -155,66 +78,197 @@ bool OddEvenRouting::is_even(int number) {
...
+            returned_legal_direction.push_back(TurnModelRouting::Direction::EAST);
+            return;
+        }
 
+        if (is_odd(comp_dst_loc.x) || diff_x > 1) {
+            returned_legal_direction.push_back(TurnModelRouting::Direction::EAST);
+        }
 
+        if (is_odd(comp_curr_loc.x) || comp_curr_loc.x == comp_src_loc.x || prev_dir != TurnModelRouting::Direction::EAST) {
+            goto route_in_yz_plane;
```
Comment (vaughnbetz) [inline vpr/src/noc/odd_even_routing.cpp:220]: Do you really need a goto?  Can't you refactor the conditional nest?

File: vpr/src/noc/west_first_routing.cpp [hunk 11]
```diff
@@ -23,105 +24,81 @@ const std::vector<TurnModelRouting::Direction>& WestFirstRouting::get_legal_dire
...
+    } else {    // 2D NoC
+        if (dst_router_pos.x < curr_router_pos.x) {
+            returned_legal_direction.push_back(TurnModelRouting::Direction::WEST);
+        } else { // to the east or the same column
+            if (dst_router_pos.x > curr_router_pos.x) { // not the same column
+                returned_legal_direction.push_back(TurnModelRouting::Direction::EAST);
+            }
+
+            if (dst_router_pos.y > curr_router_pos.y) {
+                returned_legal_direction.push_back(TurnModelRouting::Direction::UP);
```
Comment (vaughnbetz) [inline vpr/src/noc/west_first_routing.cpp:60]: NORTH

File: vpr/src/noc/west_first_routing.cpp [hunk 12]
```diff
@@ -23,105 +24,81 @@ const std::vector<TurnModelRouting::Direction>& WestFirstRouting::get_legal_dire
...
+            } else if (dst_router_pos.y < curr_router_pos.y) {
+                returned_legal_direction.push_back(TurnModelRouting::Direction::DOWN);
```
Comment (vaughnbetz) [inline vpr/src/noc/west_first_routing.cpp:62]: SOUTH

(Probably should check all UP / DOWN occurences and make sure they should not be NORTH/SOUTH)

File: vpr/src/place/initial_noc_placement.cpp [hunk 13]
```diff
@@ -179,11 +195,16 @@ static void noc_routers_anneal(const t_noc_opts& noc_opts) {
...
-    // Maximum distance in each direction that a router can travel in a move
-    // It is assumed that NoC routers are organized in a square grid.
-    // Each router can initially move within the entire grid with a single swap.
+    const auto& compressed_noc_grid = get_compressed_noc_grid();
+    const size_t n_noc_layers = compressed_noc_grid.get_layer_nums().size();
+
+    /* Maximum distance in each direction that a router can travel in a move
+     * It is assumed that NoC routers are organized in a square grid.
+     * Each router can initially move within the entire grid with a single swap.
+     * For 3D NoCs, it is assumed that each layer has the same number of routers,
```
Comment (vaughnbetz) [inline vpr/src/place/initial_noc_placement.cpp:204]: I'd clarify this: "The calculation below assumes that NoC routers are organized in a square grid; if it is a 3D architecture it also assumes each layer has the same number of routers. Breaking that assumption is OK, but the calculation may not compute the best initial range limit in that case."

**Review by vaughnbetz — CHANGES_REQUESTED** [review]
Looks good.  A few changes. Also:
- Check UP/DOWN are consistently changed to NORTH/SOUTH
- We should have at least one test for each routing algorithm (can be separate PR).

File: vpr/src/noc/odd_even_routing.h (hunk 5 above)
Comment (soheilshahrouz) [reply vpr/src/noc/odd_even_routing.h:35]: The pure virtual method in the base class (TurnModelRouting) has a doxygen comment. Should I add doxygen comments for the implementations of a pure virtual method in inheriting classes too?

Comment (vaughnbetz) [reply vpr/src/noc/odd_even_routing.h:35]: If you think the base method comment covers it then you don't need to add another one in the derived class.

**Comment by vaughnbetz** [comment]
re-launched CI. I also suggest one QoR test (Titan or VTR, could be just a circuit or two if you wish) to confirm no unexpected slowdown, and a test on the NoC designs to confirm the same.
Other than that I think this is ready to merge -- is that right?

**Comment by vaughnbetz** [comment]
Almost ready for merge ... NoC centroid move problem is resolved (was a memory bug).
Please get QoR results (2D and 3D) and we can merge this.

**Comment by soheilshahrouz** [comment]
QoR results on a subset of synthetic benchmarks on a 2D noc to measure if the introduction of 3D NoC support has slowed down NoC optimization in 2D architectures

| Metric              | Master        | 3D            | Ratio        |
|---------------------|---------------|---------------|--------------|
| Pack Time           | 333.1751111   | 339.1172979   | 1.018  |
| Place WL            | 550050.0964   | 543046.6124   | 0.987 |
| Place Time          | 372.4883381   | 365.6559439   | 0.981 |
| Total Swap          | 4183327.301   | 4189977.075   | 1.002   |
| Place CPD           | 7.164124389   | 7.170530009   | 1.001  |
| Agg BW              | 1.22E+07      | 1.22E+07      | 0.997        |
| Agg Lat             | 2.94E-07      | 2.93E-07      | 0.999        |
| Latency Overrun     | 0.00000000024 | 0.00000000024 | 1            |
| Congestion      | 0.016         | 0.016         | 1            |
| Congested Links     | 0.16          | 0.16          | 1            |

**Comment by vaughnbetz** [comment]
3D seems to have slightly improved things :).  Merging!

**Pull request merged**